cpp/src/io/utilities/data_casting.cu

```diff
@@ -0,0 +1,928 @@
+      if (!d_chars) {
+        if (lane == 0) { d_offsets[istring] = in_end - in_begin; }
+      } else {
+        for (size_type char_index = lane; char_index < (in_end - in_begin);
+             char_index += BLOCK_SIZE) {
+          d_buffer[char_index] = in_begin[char_index];
+        }
+      }
+      continue;  // gride-stride return;
+    }
+
+    // Exclude beginning and ending quote chars from string range
+    if (!options.keepquotes) {
+      ++in_begin;
+      --in_end;
+    }
+    // warp-parallelized or block-parallelized process_string()
+
+    auto is_hex = [](auto ch) {
+      return (ch >= '0' && ch <= '9') || (ch >= 'A' && ch <= 'F') || (ch >= 'a' && ch <= 'f');
+    };
+
+    // for backslash scan calculation: is_previous_escaping_backslash
+    [[maybe_unused]] auto warp_id = threadIdx.x / cudf::detail::warp_size;
+    bool init_state_reg;
+    __shared__ bool init_state_shared;
+    size_type last_offset_reg;
+    __shared__ size_type last_offset_shared;
+    bool& init_state(is_warp ? init_state_reg : init_state_shared);
+    size_type& last_offset(is_warp ? last_offset_reg : last_offset_shared);
+    if (is_warp || lane == 0) {
+      init_state  = false;
+      last_offset = 0;
+    }
+    using bitfield =
+      std::conditional_t<is_warp, bitfield_warp<num_warps>, bitfield_block<num_warps>>;
+    __shared__ bitfield is_slash;
+    is_slash.reset(warp_id);
+    __syncthreads();
+    // 0-31, 32-63, ... i*32-n.
+    // entire warp executes but with mask.
+    for (size_type char_index = lane;
+         char_index < cudf::util::round_up_safe(in_end - in_begin, static_cast<long>(BLOCK_SIZE));
+         char_index += BLOCK_SIZE) {
+      bool const is_within_bounds = char_index < (in_end - in_begin);
+      auto const MASK   = is_warp ? __ballot_sync(0xffffffff, is_within_bounds) : 0xffffffff;
+      auto const c      = is_within_bounds ? in_begin[char_index] : '\0';
+      auto const prev_c = (char_index > 0 and is_within_bounds) ? in_begin[char_index - 1] : '\0';
+      auto const escaped_char = get_escape_char(c);
+
+      bool is_escaping_backslash{false};
+      [[maybe_unused]] bool is_prev_escaping_backslash{false};
+      // To check current is backslash by checking if previous is backslash.
+      // curr = !prev & c=='\\'
+      // So, scan is required from beginning of string.
+      // State table approach (intra-warp FST) (intra-block FST)
+      // 2 states: Not-Slash(NS), Slash(S).
+      // prev  /   *
+      // NS    S  NS
+      //  S   NS  NS
+      // After inclusive scan, all current S states translate to escaping backslash.
+      // All escaping backslash should be skipped.
+
+      struct state_table {
+        bool state[2];
+      };
+      state_table curr{is_within_bounds && c == '\\', false};  // state transition vector.
+      auto composite_op = [](state_table op1, state_table op2) {
+        return state_table{op2.state[op1.state[0]], op2.state[op1.state[1]]};
+      };
```

Comment:
`cuobjdump --dump-resource-usage` shows this.
 Function void cudf::io::json::detail::parse_fn_string_parallel<(bool)0, (int)8, thrust::transform_iterator<cudf::io::json::detail::to_string_view_pair<char>, thrust::zip_iterator<thrust::tuple<const int *, const int *, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type>>, thrust::use_default, thrust::use_default>>(T3, int, int *, unsigned int *, int *, cudf::io::parse_options_view, int *, char *):
  REG:63 STACK:8 SHARED:2112 LOCAL:0 CONSTANT[0]:496 CONSTANT[2]:88 TEXTURE:0 SURFACE:0 SAMPLER:0

 Function void cudf::io::json::detail::parse_fn_string_parallel<(bool)1, (int)8, thrust::transform_iterator<cudf::io::json::detail::to_string_view_pair<char>, thrust::zip_iterator<thrust::tuple<const int *, const int *, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type>>, thrust::use_default, thrust::use_default>>(T3, int, int *, unsigned int *, int *, cudf::io::parse_options_view, int *, char *):
  REG:52 STACK:8 SHARED:296 LOCAL:0 CONSTANT[0]:496 CONSTANT[2]:88 TEXTURE:0 SURFACE:0 SAMPLER:0

The gmem-backed local memory is LOCAL right? it's zero here.

Reply:
Possibly, I can't find the documentation on `cuobjdump's` `LOCAL`. Wondering how it would index into the array if it doesn't use local memory? Does it use branching?

Reply:
Were you able to replicate the performance gains from using existing utilities?

Reply:
Another suggestion without arrays. (using bitfields in struct)
```cpp
struct state_table {
        bool state0 : 1;
        bool state1 : 1;
        // bool state[2];
      };
      state_table curr{is_within_bounds && c == '\\', false};  // state transition vector.
      auto composite_op = [](state_table op1, state_table op2) {
        // return state_table{op2.state[op1.state[0]], op2.state[op1.state[1]]};
        return state_table{ op1.state0 ? op2.state1 : op2.state0,  op1.state1 ? op2.state1 : op2.state0};
      };
```
This change reduces register and shared memory usage. Also, **eliminates STACK to zero!** 

 Function void cudf::io::json::detail::parse_fn_string_parallel<(bool)0, (int)8, thrust::transform_iterator<cudf::io::json::detail::to_string_view_pair<char>, thrust::zip_iterator<thrust::tuple<const int *, const int *, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type>>, thrust::use_default, thrust::use_default>>(T3, int, int *, unsigned int *, int *, cudf::io::parse_options_view, int *, char *):
  REG:55 STACK:0 SHARED:1824 LOCAL:0 CONSTANT[0]:496 CONSTANT[2]:88 TEXTURE:0 SURFACE:0 SAMPLER:0
 Function void cudf::io::json::detail::parse_fn_string_parallel<(bool)1, (int)8, thrust::transform_iterator<cudf::io::json::detail::to_string_view_pair<char>, thrust::zip_iterator<thrust::tuple<const int *, const int *, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type>>, thrust::use_default, thrust::use_default>>(T3, int, int *, unsigned int *, int *, cudf::io::parse_options_view, int *, char *):
  REG:50 STACK:0 SHARED:296 LOCAL:0 CONSTANT[0]:496 CONSTANT[2]:88 TEXTURE:0 SURFACE:0 SAMPLER:0

Also, this boosts speedup for block kernel, and overall speedup of -37.69% in the benchmarks! 
Using `fst::detail::MultiFragmentInRegArray` got same runtime improvements (-35.12%), but shared memory didn't reduce.

Reply:
Yup, sure, that's fine too. As long as we reap the performance gains 🙂